File: app/src/main/java/com/motrixi/datacollection/utils/GPSLocationUtil.java
package com.motrixi.datacollection.utils;

import android.Manifest;
import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.location.Address;
import android.location.Criteria;
import android.location.Geocoder;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.location.LocationProvider;
import android.os.Build;
import android.os.Bundle;
import android.provider.Settings;
import android.util.Log;
import android.widget.Toast;

import com.adobe.fre.FREContext;

import java.util.List;
import java.util.Locale;

/**
 * author : Jason
 * date   : 2020/12/8 3:34 PM
 * desc   :
 */
public class GPSLocationUtil {

    private static LocationManager mLocationManager;
    private static final String TAG = "GPSLocationUtil";
    private static Location mLocation = null;
    private static FREContext mContext;


    public static void init(FREContext context) {

        mContext = context;
        mLocationManager = (LocationManager) context.getActivity().getSystemService(Context.LOCATION_SERVICE);
        // check the gps whether on
        if (!mLocationManager.isProviderEnabled(LocationManager.GPS_PROVIDER)) {
            //Toast.makeText(context, "Please set the GPS on...", Toast.LENGTH_SHORT).show();
            return;
        }

        String bestProvider = mLocationManager.getBestProvider(getCriteria(), true);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (mContext.getActivity().checkSelfPermission(Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED
                    && mContext.getActivity().checkSelfPermission(Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {

                return;
            }
        }
        Location location = mLocationManager.getLastKnownLocation(bestProvider);
        mLocation = location;

        //mLocationManager.requestLocationUpdates(bestProvider, 2000, 0, locationListener);
        //mLocationManager.requestLocationUpdates(LocationManager.GPS_PROVIDER, 2000, 1, locationListener);
        mLocationManager.requestLocationUpdates(bestProvider, 2000, 1, locationListener);
    }

    /**
     *
     * @return
     */
    private static Criteria getCriteria() {
        Criteria criteria = new Criteria();
        //Criteria.ACCURACY_COARSE is Roughly ，Criteria.ACCURACY_FINE is Fine
        criteria.setAccuracy(Criteria.ACCURACY_FINE);
        criteria.setCostAllowed(true);
        criteria.setBearingRequired(true);
        criteria.setAltitudeRequired(true);
        criteria.setPowerRequirement(Criteria.POWER_LOW);
        return criteria;
    }

    private static LocationListener locationListener = new LocationListener() {

        //location changed
        public void onLocationChanged(Location location) {
            mLocation = location;
        }

        public void onStatusChanged(String provider, int status, Bundle extras) {
            switch (status) {
                case LocationProvider.AVAILABLE:
                    Log.d(TAG, "GPS available");
                    break;
                case LocationProvider.OUT_OF_SERVICE:
                    Log.d(TAG, "GPS out of service");
                    break;
                case LocationProvider.TEMPORARILY_UNAVAILABLE:
                    Log.d(TAG, "GPS temp unavailable");
                    break;
            }
        }

        public void onProviderEnabled(String provider) {
            @SuppressLint("MissingPermission")
            Location location = mLocationManager.getLastKnownLocation(provider);
            mLocation = location;
        }

        public void onProviderDisabled(String provider) {
            mLocation = null;
        }
    };

    /**
     * @return Location--->getLongitude()/getLatitude()
     */
    public static Location getLocation(FREContext context) {
        init(context);
        if (mLocation == null) {
            Log.e("GPSLocationUtil", "setLocationData: current location is null");
            //Toast.makeText(mContext, "current location is null", Toast.LENGTH_SHORT).show();
            return null;
        }
        return mLocation;
    }

    public static List<Address> getAddress(Location location) {
        List<Address> result = null;
        try {
            if (location != null) {
                Geocoder gc = new Geocoder(mContext.getActivity(), Locale.getDefault());
                result = gc.getFromLocation(location.getLatitude(),
                        location.getLongitude(), 1);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return result;
    }
}

File: app/src/main/java/com/motrixi/datacollection/utils/GPSLocationUtil.kt
package com.motrixi.datacollection.utils

import android.Manifest
import android.annotation.SuppressLint
import android.content.Context
import android.content.pm.PackageManager
import android.location.*
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.widget.Toast
import java.util.*

/**
 * author : Jason
 *  date   : 2020/10/21 2:13 PM
 *  desc   :
 */
object GPSLocationUtil {

    private var mLocationManager: LocationManager? = null
    private const val TAG = "GPSLocationUtil"
    private var mLocation: Location? = null
    private var mContext: Context? = null

    fun init(context: Context) {
        mContext = context
        mLocationManager = context.getSystemService(Context.LOCATION_SERVICE) as LocationManager

        // check the gps whether on
        if (!mLocationManager!!.isProviderEnabled(LocationManager.GPS_PROVIDER)) {
            Toast.makeText(context, "Please set the GPS on...", Toast.LENGTH_SHORT).show()
            //val intent = Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS)
            //mContext.startActivityForResult(intent, 0)
            return
        }
        val bestProvider = mLocationManager!!.getBestProvider(getCriteria()!!, true)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (mContext!!.checkSelfPermission(Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED
                && mContext!!.checkSelfPermission(Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED
            ) {
                return
            }
        }
        val location = mLocationManager!!.getLastKnownLocation(bestProvider!!)
        mLocation = location
        //mLocationManager!!.addGpsStatusListener(listener)

        //mLocationManager.requestLocationUpdates(bestProvider, 2000, 0, locationListener);
        //mLocationManager.requestLocationUpdates(LocationManager.GPS_PROVIDER, 2000, 1, locationListener);
        mLocationManager!!.requestLocationUpdates(bestProvider, 2000, 1f, locationListener)
    }

    /**
     *
     * @return
     */
    private fun getCriteria(): Criteria? {
        val criteria = Criteria()
        //Criteria.ACCURACY_COARSE is Roughly ，Criteria.ACCURACY_FINE is Fine
        criteria.accuracy = Criteria.ACCURACY_FINE
        criteria.isCostAllowed = true
        criteria.isBearingRequired = true
        criteria.isAltitudeRequired = true
        criteria.powerRequirement = Criteria.POWER_LOW
        return criteria
    }

    private val locationListener: LocationListener = object : LocationListener {
        //location changed
        override fun onLocationChanged(location: Location) {
            mLocation = location
            if (mLocation != null) {
                Log.d(TAG, "time：" + location.time)
                Log.d(TAG, "lon：" + location.longitude)
                Log.d(TAG, "lat：" + location.latitude)
            }

        }

        override fun onStatusChanged(provider: String, status: Int, extras: Bundle) {
            when (status) {
                LocationProvider.AVAILABLE -> Log.d(TAG, "GPS available")
                LocationProvider.OUT_OF_SERVICE -> Log.d(TAG, "GPS out of service")
                LocationProvider.TEMPORARILY_UNAVAILABLE -> Log.d(TAG, "GPS temp unavailable")
            }
        }

        override fun onProviderEnabled(provider: String) {
            @SuppressLint("MissingPermission")
            val location = mLocationManager!!.getLastKnownLocation(provider)
            mLocation = location
        }

        override fun onProviderDisabled(provider: String) {
            mLocation = null
        }
    }

    /**
     * @return Location--->getLongitude()/getLatitude()
     */
    fun getLocation(context: Context): Location? {
        init(context)
        if (mLocation == null) {
            Log.d("GPSLocationUtil", "setLocationData: current location is null")
            //Toast.makeText(mContext, "current location is null", Toast.LENGTH_SHORT).show()
            return null
        }
        return mLocation
    }

    fun getLocalCity(): String? {
        if (mLocation == null) {
            Log.d(TAG, "getLocalCity: city info is null")
            return ""
        }
        val result = getAddress(mLocation)
        var city: String? = ""
        if (result != null && result.size > 0) {
            city = result[0].locality //get city info
        }
        return city
    }

    fun getAddressStr(): String? {
        if (mLocation == null) {
            Log.d(TAG, "getAddressStr: address info is null")
            return ""
        }
        val result = getAddress(mLocation)
        var address: String? = ""
        if (result != null && result.size > 0) {
            address = result[0].getAddressLine(0) //get address info
        }
        return address
    }

    fun getAddress(location: Location?): List<Address>? {
        var result: List<Address>? = null
        try {
            if (location != null) {
                val gc = Geocoder(mContext, Locale.getDefault())
//                val gc = Geocoder(mContext, Locale.ENGLISH)
                result = gc.getFromLocation(location.latitude, location.longitude, 1)
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return result
    }


    /*@SuppressLint("MissingPermission")
    var listener = GpsStatus.Listener { event ->
        when (event) {
            GpsStatus.GPS_EVENT_FIRST_FIX -> Log.d(TAG, "first location")
            GpsStatus.GPS_EVENT_SATELLITE_STATUS -> {
                val gpsStatus = mLocationManager!!.getGpsStatus(null)
                val maxSatellites = gpsStatus!!.maxSatellites
                val iters: Iterator<GpsSatellite> = gpsStatus.satellites.iterator()
                var count = 0
                while (iters.hasNext() && count <= maxSatellites) {
                    val s = iters.next()
                    count++
                }
                //Log.d(TAG, "satellite count: $count")
            }
            GpsStatus.GPS_EVENT_STARTED -> Log.d(TAG, "location start")
            GpsStatus.GPS_EVENT_STOPPED -> Log.d(TAG, "location stop")
        }
    }*/
}